### PR TITLE
EDGECLOUD-2001 Add object detection using GPU-accelerated library

### PR DESCRIPTION
**Pull request by brucearmstrong**
- Uses pytorch library for object detection. Uses GPU if available, otherwise CPU only.
- Added pytorch requirements.
- All endpoints now use common get_image_from_request function, so they now all support multiple content_type values, including those produced by "curl -F".
- Additions to server_tester.py and tcp_client.py to add up and average the server_processing_time values from the server.

To avoid checking in code that didn't have a license included in its repository, we are now using submodules (https://git-scm.com/book/en/v2/Git-Tools-Submodules) to pull in the entire project. The new object_detector.py adds the submodule's directory to the path, then uses the packages as usual.

### CODE REVIEW

File: FaceDetectionServer/moedx/facial_detection/tcp_server.py [hunk 1]
```diff
@@ -100,6 +96,21 @@ def handle(self):
                         num_poses = len(poses2)
                         ret = {"success": "true", "server_processing_time": elapsed, "poses": poses2.tolist()}
 
+            elif opcode == 4: #'obj_det':
+                now = time.time()
+                image = imread(io.BytesIO(data))
+                pillow_image = Image.fromarray(image, 'RGB')
+                objects = myObjectDetector.process_image(pillow_image)
+                elapsed = "%.3f" %((time.time() - now)*1000)
```
Comment (franklin-huang-mobiledgex) [inline FaceDetectionServer/moedx/facial_detection/tcp_server.py]: What does "%.3f" do?

Comment (brucearmstrong) [reply FaceDetectionServer/moedx/facial_detection/tcp_server.py]: Formats a floating point number to 3 decimal places.

File: FaceDetectionServer/moedx/object_detection/object_detector.py [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+        pass
+
+    def detect_image(self, img):
+        # scale and pad image
+        ratio = min(img_size/img.size[0], img_size/img.size[1])
+        imw = round(img.size[0] * ratio)
+        imh = round(img.size[1] * ratio)
+        img_transforms = transforms.Compose([ transforms.Resize((imh, imw)),
+             transforms.Pad((max(int((imh-imw)/2),0), max(int((imw-imh)/2),0), max(int((imh-imw)/2),0), max(int((imw-imh)/2),0)),
+                            (128,128,128)),
```
Comment (franklin-huang-mobiledgex) [inline FaceDetectionServer/moedx/object_detection/object_detector.py]: What is (128, 128, 128)? A color?

File: FaceDetectionServer/moedx/object_detection/object_detector.py [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+            colors = [cmap(i) for i in np.linspace(0, 1, 20)]
+
+            plt.figure()
+            fig, ax = plt.subplots(1, figsize=(12,9))
+            ax.imshow(img)
+        else:
+            img = np.array(img)
+
+        pad_x = max(img.shape[0] - img.shape[1], 0) * (img_size / max(img.shape))
+        pad_y = max(img.shape[1] - img.shape[0], 0) * (img_size / max(img.shape))
```
Comment (franklin-huang-mobiledgex) [inline FaceDetectionServer/moedx/object_detection/object_detector.py]: I'm a little bit confused by the padding logic (also in line 69). Are we subtracting the height from the width and vice versa? Is this to make the image a square?

File: FaceDetectionServer/moedx/object_detection/object_detector.py (hunk 2 above)
Comment (brucearmstrong) [reply FaceDetectionServer/moedx/object_detection/object_detector.py]: Yes. When the image is resized, it builds a gray border around it to maintain aspect ratio of the original image.

File: FaceDetectionServer/moedx/object_detection/object_detector.py (hunk 3 above)
Comment (brucearmstrong) [reply FaceDetectionServer/moedx/object_detection/object_detector.py]: Yes, the image must be a square, but with gray padding to maintain aspect ration. Picture a letterboxed or pillarboxed movie.

**Pull request merged**